### PR TITLE
feat(router): schema cache anti-entropy/sync actor

### PR DESCRIPTION
**Pull request by domodwyer**
This PR builds on #8780, consuming the merkle search tree content and eventually using it to drive eventual consistency of the schema cache content across gossip peers.

This PR introduces the client-side sync mechanism driving eventual consistency of the schema cache content across gossip peers, using the sync protocol described in https://github.com/influxdata/influxdb_iox/pull/8753.

A follow-up PR will introduce the server-side component this client interacts with.

---

* refactor: ColumnsByName default impl (9493ce74a)
      
      The ColumnsByName type can/should default to an empty set.

* feat(router): schema cache anti-entropy/sync actor (ee0334dab)
      
      Adds the foundational components of the router anti-entropy subsystem.
      This subsystem compliments the existing schema gossiping mechanism to
      provide eventual consistency of the router schema caches.
      
      This commit introduces the "convergence actor", an actor task
      responsible for detecting peer inconsistencies and efficiently
      converging those differences using the existing Merkle Search Tree
      actor.
      
      Various sub-components break this convergence problem down into simple,
      singly-responsible parts:
      
        * The existing Merkle Search Tree actor tracking the NamespaceCache
          content
      
        * A consistency prober to send MST root hashes to peers over gossip
      
        * An actor task to send and and process received consistency probes
          from peers, and start a sync worker to converge inconsistencies
      
        * A sync worker that drives MST differential syncing and convergence
          against a single peer
      
      The sync worker makes use of the existing gossip message format used to
      communicate schema cache changes between peers, and drives convergence
      by fetching & replaying the messages necessary to fully construct the
      diverged schemas; reusing the gossip messages ensures all schema updates
      flow through the same code paths, regardless of source (sync vs gossip).

### CODE REVIEW

File: router/src/gossip/anti_entropy/sync/consistency_prober.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        let payload = ConsistencyProbe {
+            root_hash: local_root.as_bytes().to_vec(),
+            grpc_port: rpc_bind as _,
+        };
+
+        // Broadcast the root hash to a random subset of cluster peers.
+        self.broadcast_subset(payload.encode_to_vec(), Topic::SchemaCacheConsistency)
+            .await
+            .expect("gossip message size exceeds max");
+    }
```
Comment (domodwyer) [inline router/src/gossip/anti_entropy/sync/consistency_prober.rs:36]: The default `ConsistencyProber` implementation simply sends the merkle tree root hash to a random subset of gossip peers.

File: router/src/gossip/anti_entropy/sync/rpc_worker.rs [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+                for table in event.tables {
+                    self.schema_event_delegate
+                        .handle(Event::TableCreated(table))
+                        .await;
+                }
+            }
+        }
+
+        Ok(())
+    }
```
Comment (domodwyer) [inline router/src/gossip/anti_entropy/sync/rpc_worker.rs:135]: An the sync `RpcWorker` is relatively simple too - it just calls the two RPC methods on the inconsistent peer, using it to perform a diff and then pull the identified inconsistent pages to converge.

File: router/src/gossip/anti_entropy/sync/rpc_worker.rs [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+        let sync_worker = RpcWorker::new(
+            Arc::clone(&rpc_client),
+            schema_event_handler,
+            mst,
+            MOCK_ADDR,
+            Identity::try_from(b"1234567890123456".to_vec()).unwrap(),
+        );
+
+        // Run the sync worker to completion.
+        sync_worker.run().await;
```
Comment (savage-engineer) [inline router/src/gossip/anti_entropy/sync/rpc_worker.rs:284]: Could be worth checking the contents of the `MockSyncRpcClient`s call list here before getting into assertions that throw up diffs. May be a bit on the boiler plate side, YMMV.

File: router/src/gossip/anti_entropy/sync/consistency_prober.rs [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        // Generate the solicitation payload.
+        let payload = ConsistencyProbe {
+            root_hash: local_root.as_bytes().to_vec(),
+            grpc_port: rpc_bind as _,
+        };
+
+        // Broadcast the root hash to a random subset of cluster peers.
+        self.broadcast_subset(payload.encode_to_vec(), Topic::SchemaCacheConsistency)
+            .await
+            .expect("gossip message size exceeds max");
```
Comment (savage-engineer) [inline router/src/gossip/anti_entropy/sync/consistency_prober.rs:35]: Presumably this `expect` call is safe (without arbitrary fields being wanged into the message definition) as the root hash is fixed size at 16 bytes?

**Review by savage-engineer — APPROVED** [review]

File: router/src/gossip/anti_entropy/sync/consistency_prober.rs (hunk 4 above)
Comment (domodwyer) [reply router/src/gossip/anti_entropy/sync/consistency_prober.rs:35]: Correct - there'll be a test later too :+1:

**Pull request merged**